lib/wallet/src/Cardano/Wallet/Write/Tx.hs

```diff
@@ -491,9 +619,74 @@ utxoFromTxOutsInLatestEra
 utxoFromTxOutsInLatestEra = withStandardCryptoConstraint RecentEraBabbage $
     Shelley.UTxO . Map.fromList . map (second castTxOut)
 
+--------------------------------------------------------------------------------
+-- Tx
+--------------------------------------------------------------------------------
+
+modifyTxOutputs
+    :: RecentEra era
+    -> (TxOut (ShelleyLedgerEra era) -> TxOut (ShelleyLedgerEra era))
+    -> Core.TxBody (ShelleyLedgerEra era)
+    -> Core.TxBody (ShelleyLedgerEra era)
+modifyTxOutputs era f body = case era of
+    RecentEraBabbage -> body
+        { Babbage.outputs = mapSized f <$> Babbage.outputs body
+        }
+    RecentEraAlonzo -> body
+        { Alonzo.outputs = f <$> Alonzo.outputs body
+        }
+  where
+    mapSized f' = mkSized . f' . sizedValue
+
+txBody
+    :: RecentEra era
+    -> Core.Tx (ShelleyLedgerEra era)
+    -> Core.TxBody (ShelleyLedgerEra era)
+txBody RecentEraBabbage = Alonzo.body -- same type for babbage
+txBody RecentEraAlonzo = Alonzo.body
+
+-- Until we have convenient lenses to use
+outputs
+    :: RecentEra era
+    -> Core.TxBody (ShelleyLedgerEra era)
+    -> [TxOut (ShelleyLedgerEra era)]
+outputs RecentEraBabbage = map sizedValue . toList . Babbage.outputs
+outputs RecentEraAlonzo = toList . Alonzo.outputs
+
+-- NOTE: To reduce the need for the caller to deal with @CardanoApiEra
+-- (ShelleyLedgerEra era) ~ era@, we quantify this function over @cardanoEra@
+-- instead of @era@.
+--
+-- TODO [ADP-2353] Move to @cardano-api@ related module
+modifyLedgerBody
+    :: (Core.TxBody (ShelleyLedgerEra cardanoEra)
+        -> Core.TxBody (ShelleyLedgerEra cardanoEra))
+    -> Cardano.Tx cardanoEra
+    -> Cardano.Tx cardanoEra
+modifyLedgerBody f (Cardano.Tx body keyWits) =
+    let
+        Cardano.ShelleyTxBody
+            shelleyEra
+            ledgerBody
+            scripts
+            scriptData
+            auxData
+            validity
+            = body
+        body' = Cardano.ShelleyTxBody
+            shelleyEra
+            (f ledgerBody)
+            scripts
+            scriptData
+            auxData
+            validity
+    in
+        Cardano.Tx body' keyWits
```

Comment:
The GADTs defined by the ledger are nice, but a real pain if we just want to update one field!

How nice it would be if we could write something like this:

```hs
modifyLedgerBody = over (#body . #ledgerBody)
```

Reply:
I was just thinking yesterday about record dot syntax and how `HasField` constraints could be added as superclass to `RecentEra` for the ultimate ease-of-use.

But you're right, something similar could probably be done today already...

Thanks for reminding me about `over` — I'll have a look!

Reply:
Tentatively I think we'd need to define e.g.
```haskell
instance Data.Generics.Product.Fields.HasField' "value" (Babbage.TxOut StandardBabbage) (Mary.Value StandardCrypto) where
```
instance to make generic-lens work.

On our current revision of the ledger there are a lot of `GHC.Records.HasField` instances which don't include setters.

**However**, on more recent `cardano-ledger` revisions it seems they have added `Generic` instances to types. I think those generic instances together with turning `IsRecentEra era` into `class ManyHasFieldConstraints => IsRecentEra era where` would be what we'd need to remove the helper functions.

Reply:
I suspect we should not let the helpers block the merge of this PR, but it would be **very** nice to get this kind of thing working!